apm-agent-core/src/main/java/co/elastic/apm/agent/metrics/Labels.java

```diff
@@ -0,0 +1,380 @@
+                labels.add(entry.getKey(), entry.getValue());
+            }
+            return labels;
+        }
+
+        public Labels add(String key, CharSequence value) {
+            keys.add(key);
+            values.add(value);
+            return this;
+        }
+
+        public Labels.Mutable transactionName(@Nullable CharSequence transactionName) {
+            this.transactionName = transactionName;
+            return this;
+        }
+
+        public Labels.Mutable transactionType(@Nullable String transactionType) {
+            this.transactionType = transactionType;
+            return this;
+        }
+
+        public Labels.Mutable spanType(@Nullable String spanType) {
+            this.spanType = spanType;
+            return this;
+        }
+
+        public Labels.Mutable spanSubType(@Nullable String subtype) {
+            this.spanSubType = subtype;
+            return this;
+        }
+
+        @Nullable
+        public CharSequence getTransactionName() {
+            return transactionName;
+        }
+
+        @Nullable
+        public String getTransactionType() {
+            return transactionType;
+        }
+
+        @Nullable
+        public String getSpanType() {
+            return spanType;
+        }
+
+        @Override
+        @Nullable
+        public String getSpanSubType() {
+            return spanSubType;
+        }
+
+        public Labels.Immutable immutableCopy() {
+            return new Immutable(this);
+        }
+
+        @Override
+        public void resetState() {
+            keys.clear();
+            values.clear();
+            transactionName = null;
+            transactionType = null;
+            spanType = null;
+        }
+    }
+
+    /**
+     * An immutable implementation of the {@link Labels} interface
+     * <p>
+     * To publish a copy of {@link Mutable} in a thread-safe manner,
+     * all properties need to be final.
+     * That's why we can't share the exact same class.
+     * </p>
+     */
+    class Immutable extends AbstractBase {
+        private static final Labels.Immutable EMPTY = new Mutable().immutableCopy();
+
+        private final int hash;
+        @Nullable
+        private final String transactionName;
+        @Nullable
+        private final String transactionType;
+        @Nullable
+        private final String spanType;
+        @Nullable
+        private final String spanSubType;
+
+        public Immutable(Labels labels) {
+            super(new ArrayList<>(labels.getKeys()), copy(labels.getValues()));
+            final CharSequence transactionName = labels.getTransactionName();
+            this.transactionName = transactionName != null ? transactionName.toString() : null;
+            this.transactionType = labels.getTransactionType();
+            this.spanType = labels.getSpanType();
+            this.spanSubType = labels.getSpanSubType();
+            this.hash = labels.hashCode();
```

Comment:
could you explain?

Reply:
You are constructing an Immutable probably from a Mutable object, which may change during construction, for example (if used concurrently). Maybe in the future we will add some manipulation to the `labels` fields before assigning them in the Immutable, so this will also make it invalid.
Even if you can't see a scenario where this may currently actually break, it seems more correct to calculate an Object's hash code based on its own fields, and not another Object's fields.

Reply:
Copying is not thread-safe in either case - if there are labels added concurrently to creating an immutable copy there can be inconsistencies. Bottom line is that `Labels.Mutable` is not thread-safe while `Labels.Immutable` is.

Reply:
> Copying is not thread-safe in either case - if there are labels added concurrently to creating an immutable copy there can be inconsistencies

This is not about maintaining consistency between the original and copy- you can't guarantee consistency between Mutable and Immutable, regardless of concurrency (the mutable may change well after copying). 

I demonstrated a case where the `hashCode` of the Object would not reflect it's fields- if after you set `this.spanType = labels.getSpanType()`, then other thread changes `labels#spanType`, then you set `this.hash = labels.hashCode()` - this is what you'd get.

Bottom line- more correct to store an Immutable's final `hashCode` based on its own final fields.